File: src/protobuf/python/google/protobuf/internal/wire_format_test.py
#! /usr/bin/python
#
# protocol buffers - google's data interchange format
# copyright 2008 google inc.  all rights reserved.
# http://code.google.com/p/protobuf/
#
# redistribution and use in source and binary forms, with or without
# modification, are permitted provided that the following conditions are
# met:
#
#     * redistributions of source code must retain the above copyright
# notice, this list of conditions and the following disclaimer.
#     * redistributions in binary form must reproduce the above
# copyright notice, this list of conditions and the following disclaimer
# in the documentation and/or other materials provided with the
# distribution.
#     * neither the name of google inc. nor the names of its
# contributors may be used to endorse or promote products derived from
# this software without specific prior written permission.
#
# this software is provided by the copyright holders and contributors
# "as is" and any express or implied warranties, including, but not
# limited to, the implied warranties of merchantability and fitness for
# a particular purpose are disclaimed. in no event shall the copyright
# owner or contributors be liable for any direct, indirect, incidental,
# special, exemplary, or consequential damages (including, but not
# limited to, procurement of substitute goods or services; loss of use,
# data, or profits; or business interruption) however caused and on any
# theory of liability, whether in contract, strict liability, or tort
# (including negligence or otherwise) arising in any way out of the use
# of this software, even if advised of the possibility of such damage.

"""test for google.protobuf.internal.wire_format."""

__author__ = 'robinson@google.com (will robinson)'

import unittest
from google.protobuf import message
from google.protobuf.internal import wire_format


class wireformattest(unittest.testcase):

  def testpacktag(self):
    field_number = 0xabc
    tag_type = 2
    self.assertequal((field_number << 3) | tag_type,
                     wire_format.packtag(field_number, tag_type))
    packtag = wire_format.packtag
    # number too high.
    self.assertraises(message.encodeerror, packtag, field_number, 6)
    # number too low.
    self.assertraises(message.encodeerror, packtag, field_number, -1)

  def testunpacktag(self):
    # test field numbers that will require various varint sizes.
    for expected_field_number in (1, 15, 16, 2047, 2048):
      for expected_wire_type in range(6):  # highest-numbered wiretype is 5.
        field_number, wire_type = wire_format.unpacktag(
            wire_format.packtag(expected_field_number, expected_wire_type))
        self.assertequal(expected_field_number, field_number)
        self.assertequal(expected_wire_type, wire_type)

    self.assertraises(typeerror, wire_format.unpacktag, none)
    self.assertraises(typeerror, wire_format.unpacktag, 'abc')
    self.assertraises(typeerror, wire_format.unpacktag, 0.0)
    self.assertraises(typeerror, wire_format.unpacktag, object())

  def testzigzagencode(self):
    z = wire_format.zigzagencode
    self.assertequal(0, z(0))
    self.assertequal(1, z(-1))
    self.assertequal(2, z(1))
    self.assertequal(3, z(-2))
    self.assertequal(4, z(2))
    self.assertequal(0xfffffffe, z(0x7fffffff))
    self.assertequal(0xffffffff, z(-0x80000000))
    self.assertequal(0xfffffffffffffffe, z(0x7fffffffffffffff))
    self.assertequal(0xffffffffffffffff, z(-0x8000000000000000))

    self.assertraises(typeerror, z, none)
    self.assertraises(typeerror, z, 'abcd')
    self.assertraises(typeerror, z, 0.0)
    self.assertraises(typeerror, z, object())

  def testzigzagdecode(self):
    z = wire_format.zigzagdecode
    self.assertequal(0, z(0))
    self.assertequal(-1, z(1))
    self.assertequal(1, z(2))
    self.assertequal(-2, z(3))
    self.assertequal(2, z(4))
    self.assertequal(0x7fffffff, z(0xfffffffe))
    self.assertequal(-0x80000000, z(0xffffffff))
    self.assertequal(0x7fffffffffffffff, z(0xfffffffffffffffe))
    self.assertequal(-0x8000000000000000, z(0xffffffffffffffff))

    self.assertraises(typeerror, z, none)
    self.assertraises(typeerror, z, 'abcd')
    self.assertraises(typeerror, z, 0.0)
    self.assertraises(typeerror, z, object())

  def numericbytesizetesthelper(self, byte_size_fn, value, expected_value_size):
    # use field numbers that cause various byte sizes for the tag information.
    for field_number, tag_bytes in ((15, 1), (16, 2), (2047, 2), (2048, 3)):
      expected_size = expected_value_size + tag_bytes
      actual_size = byte_size_fn(field_number, value)
      self.assertequal(expected_size, actual_size,
                       'byte_size_fn: %s, field_number: %d, value: %r\n'
                       'expected: %d, actual: %d'% (
          byte_size_fn, field_number, value, expected_size, actual_size))

  def testbytesizefunctions(self):
    # test all numeric *bytesize() functions.
    numeric_args = [
        # int32bytesize().
        [wire_format.int32bytesize, 0, 1],
        [wire_format.int32bytesize, 127, 1],
        [wire_format.int32bytesize, 128, 2],
        [wire_format.int32bytesize, -1, 10],
        # int64bytesize().
        [wire_format.int64bytesize, 0, 1],
        [wire_format.int64bytesize, 127, 1],
        [wire_format.int64bytesize, 128, 2],
        [wire_format.int64bytesize, -1, 10],
        # uint32bytesize().
        [wire_format.uint32bytesize, 0, 1],
        [wire_format.uint32bytesize, 127, 1],
        [wire_format.uint32bytesize, 128, 2],
        [wire_format.uint32bytesize, wire_format.uint32_max, 5],
        # uint64bytesize().
        [wire_format.uint64bytesize, 0, 1],
        [wire_format.uint64bytesize, 127, 1],
        [wire_format.uint64bytesize, 128, 2],
        [wire_format.uint64bytesize, wire_format.uint64_max, 10],
        # sint32bytesize().
        [wire_format.sint32bytesize, 0, 1],
        [wire_format.sint32bytesize, -1, 1],
        [wire_format.sint32bytesize, 1, 1],
        [wire_format.sint32bytesize, -63, 1],
        [wire_format.sint32bytesize, 63, 1],
        [wire_format.sint32bytesize, -64, 1],
        [wire_format.sint32bytesize, 64, 2],
        # sint64bytesize().
        [wire_format.sint64bytesize, 0, 1],
        [wire_format.sint64bytesize, -1, 1],
        [wire_format.sint64bytesize, 1, 1],
        [wire_format.sint64bytesize, -63, 1],
        [wire_format.sint64bytesize, 63, 1],
        [wire_format.sint64bytesize, -64, 1],
        [wire_format.sint64bytesize, 64, 2],
        # fixed32bytesize().
        [wire_format.fixed32bytesize, 0, 4],
        [wire_format.fixed32bytesize, wire_format.uint32_max, 4],
        # fixed64bytesize().
        [wire_format.fixed64bytesize, 0, 8],
        [wire_format.fixed64bytesize, wire_format.uint64_max, 8],
        # sfixed32bytesize().
        [wire_format.sfixed32bytesize, 0, 4],
        [wire_format.sfixed32bytesize, wire_format.int32_min, 4],
        [wire_format.sfixed32bytesize, wire_format.int32_max, 4],
        # sfixed64bytesize().
        [wire_format.sfixed64bytesize, 0, 8],
        [wire_format.sfixed64bytesize, wire_format.int64_min, 8],
        [wire_format.sfixed64bytesize, wire_format.int64_max, 8],
        # floatbytesize().
        [wire_format.floatbytesize, 0.0, 4],
        [wire_format.floatbytesize, 1000000000.0, 4],
        [wire_format.floatbytesize, -1000000000.0, 4],
        # doublebytesize().
        [wire_format.doublebytesize, 0.0, 8],
        [wire_format.doublebytesize, 1000000000.0, 8],
        [wire_format.doublebytesize, -1000000000.0, 8],
        # boolbytesize().
        [wire_format.boolbytesize, false, 1],
        [wire_format.boolbytesize, true, 1],
        # enumbytesize().
        [wire_format.enumbytesize, 0, 1],
        [wire_format.enumbytesize, 127, 1],
        [wire_format.enumbytesize, 128, 2],
        [wire_format.enumbytesize, wire_format.uint32_max, 5],
        ]
    for args in numeric_args:
      self.numericbytesizetesthelper(*args)

    # test strings and bytes.
    for byte_size_fn in (wire_format.stringbytesize, wire_format.bytesbytesize):
      # 1 byte for tag, 1 byte for length, 3 bytes for contents.
      self.assertequal(5, byte_size_fn(10, 'abc'))
      # 2 bytes for tag, 1 byte for length, 3 bytes for contents.
      self.assertequal(6, byte_size_fn(16, 'abc'))
      # 2 bytes for tag, 2 bytes for length, 128 bytes for contents.
      self.assertequal(132, byte_size_fn(16, 'a' * 128))

    # test utf-8 string byte size calculation.
    # 1 byte for tag, 1 byte for length, 8 bytes for content.
    self.assertequal(10, wire_format.stringbytesize(
        5, unicode('\xd0\xa2\xd0\xb5\xd1\x81\xd1\x82', 'utf-8')))

    class mockmessage(object):
      def __init__(self, byte_size):
        self.byte_size = byte_size
      def bytesize(self):
        return self.byte_size

    message_byte_size = 10
    mock_message = mockmessage(byte_size=message_byte_size)
    # test groups.
    # (2 * 1) bytes for begin and end tags, plus message_byte_size.
    self.assertequal(2 + message_byte_size,
                     wire_format.groupbytesize(1, mock_message))
    # (2 * 2) bytes for begin and end tags, plus message_byte_size.
    self.assertequal(4 + message_byte_size,
                     wire_format.groupbytesize(16, mock_message))

    # test messages.
    # 1 byte for tag, plus 1 byte for length, plus contents.
    self.assertequal(2 + mock_message.byte_size,
                     wire_format.messagebytesize(1, mock_message))
    # 2 bytes for tag, plus 1 byte for length, plus contents.
    self.assertequal(3 + mock_message.byte_size,
                     wire_format.messagebytesize(16, mock_message))
    # 2 bytes for tag, plus 2 bytes for length, plus contents.
    mock_message.byte_size = 128
    self.assertequal(4 + mock_message.byte_size,
                     wire_format.messagebytesize(16, mock_message))


    # test message set item byte size.
    # 4 bytes for tags, plus 1 byte for length, plus 1 byte for type_id,
    # plus contents.
    mock_message.byte_size = 10
    self.assertequal(mock_message.byte_size + 6,
                     wire_format.messagesetitembytesize(1, mock_message))

    # 4 bytes for tags, plus 2 bytes for length, plus 1 byte for type_id,
    # plus contents.
    mock_message.byte_size = 128
    self.assertequal(mock_message.byte_size + 7,
                     wire_format.messagesetitembytesize(1, mock_message))

    # 4 bytes for tags, plus 2 bytes for length, plus 2 byte for type_id,
    # plus contents.
    self.assertequal(mock_message.byte_size + 8,
                     wire_format.messagesetitembytesize(128, mock_message))

    # too-long varint.
    self.assertraises(message.encodeerror,
                      wire_format.uint64bytesize, 1, 1 << 128)


if __name__ == '__main__':
  unittest.main()

File: src/protobuf/java/src/test/java/com/google/protobuf/WireFormatTest.java
// protocol buffers - google's data interchange format
// copyright 2008 google inc.  all rights reserved.
// http://code.google.com/p/protobuf/
//
// redistribution and use in source and binary forms, with or without
// modification, are permitted provided that the following conditions are
// met:
//
//     * redistributions of source code must retain the above copyright
// notice, this list of conditions and the following disclaimer.
//     * redistributions in binary form must reproduce the above
// copyright notice, this list of conditions and the following disclaimer
// in the documentation and/or other materials provided with the
// distribution.
//     * neither the name of google inc. nor the names of its
// contributors may be used to endorse or promote products derived from
// this software without specific prior written permission.
//
// this software is provided by the copyright holders and contributors
// "as is" and any express or implied warranties, including, but not
// limited to, the implied warranties of merchantability and fitness for
// a particular purpose are disclaimed. in no event shall the copyright
// owner or contributors be liable for any direct, indirect, incidental,
// special, exemplary, or consequential damages (including, but not
// limited to, procurement of substitute goods or services; loss of use,
// data, or profits; or business interruption) however caused and on any
// theory of liability, whether in contract, strict liability, or tort
// (including negligence or otherwise) arising in any way out of the use
// of this software, even if advised of the possibility of such damage.

package com.google.protobuf;

import junit.framework.testcase;

import java.io.bytearrayinputstream;
import java.io.bytearrayoutputstream;
import java.util.list;

import protobuf_unittest.unittestproto;
import protobuf_unittest.unittestproto.testallextensions;
import protobuf_unittest.unittestproto.testalltypes;
import protobuf_unittest.unittestproto.testfieldorderings;
import protobuf_unittest.unittestproto.testpackedextensions;
import protobuf_unittest.unittestproto.testpackedtypes;
import protobuf_unittest.unittestmset.testmessageset;
import protobuf_unittest.unittestmset.rawmessageset;
import protobuf_unittest.unittestmset.testmessagesetextension1;
import protobuf_unittest.unittestmset.testmessagesetextension2;
import com.google.protobuf.unittestlite.testallextensionslite;
import com.google.protobuf.unittestlite.testpackedextensionslite;

/**
 * tests related to parsing and serialization.
 *
 * @author kenton@google.com (kenton varda)
 */
public class wireformattest extends testcase {
  public void testserialization() throws exception {
    testalltypes message = testutil.getallset();

    bytestring rawbytes = message.tobytestring();
    assertequals(rawbytes.size(), message.getserializedsize());

    testalltypes message2 = testalltypes.parsefrom(rawbytes);

    testutil.assertallfieldsset(message2);
  }

  public void testserializationpacked() throws exception {
    testpackedtypes message = testutil.getpackedset();

    bytestring rawbytes = message.tobytestring();
    assertequals(rawbytes.size(), message.getserializedsize());

    testpackedtypes message2 = testpackedtypes.parsefrom(rawbytes);

    testutil.assertpackedfieldsset(message2);
  }

  public void testserializeextensions() throws exception {
    // testalltypes and testallextensions should have compatible wire formats,
    // so if we serialize a testallextensions then parse it as testalltypes
    // it should work.

    testallextensions message = testutil.getallextensionsset();
    bytestring rawbytes = message.tobytestring();
    assertequals(rawbytes.size(), message.getserializedsize());

    testalltypes message2 = testalltypes.parsefrom(rawbytes);

    testutil.assertallfieldsset(message2);
  }

  public void testserializepackedextensions() throws exception {
    // testpackedtypes and testpackedextensions should have compatible wire
    // formats; check that they serialize to the same string.
    testpackedextensions message = testutil.getpackedextensionsset();
    bytestring rawbytes = message.tobytestring();

    testpackedtypes message2 = testutil.getpackedset();
    bytestring rawbytes2 = message2.tobytestring();

    assertequals(rawbytes, rawbytes2);
  }

  public void testserializationpackedwithoutgetserializedsize()
      throws exception {
    // write directly to an outputstream, without invoking getserializedsize()
    // this used to be a bug where the size of a packed field was incorrect,
    // since getserializedsize() was never invoked.
    testpackedtypes message = testutil.getpackedset();

    // directly construct a codedoutputstream around the actual outputstream,
    // in case writeto(outputstream output) invokes getserializedsize();
    bytearrayoutputstream outputstream = new bytearrayoutputstream();
    codedoutputstream codedoutput = codedoutputstream.newinstance(outputstream);

    message.writeto(codedoutput);

    codedoutput.flush();

    testpackedtypes message2 = testpackedtypes.parsefrom(
        outputstream.tobytearray());

    testutil.assertpackedfieldsset(message2);
  }

  public void testserializeextensionslite() throws exception {
    // testalltypes and testallextensions should have compatible wire formats,
    // so if we serialize a testallextensions then parse it as testalltypes
    // it should work.

    testallextensionslite message = testutil.getallliteextensionsset();
    bytestring rawbytes = message.tobytestring();
    assertequals(rawbytes.size(), message.getserializedsize());

    testalltypes message2 = testalltypes.parsefrom(rawbytes);

    testutil.assertallfieldsset(message2);
  }

  public void testserializepackedextensionslite() throws exception {
    // testpackedtypes and testpackedextensions should have compatible wire
    // formats; check that they serialize to the same string.
    testpackedextensionslite message = testutil.getlitepackedextensionsset();
    bytestring rawbytes = message.tobytestring();

    testpackedtypes message2 = testutil.getpackedset();
    bytestring rawbytes2 = message2.tobytestring();

    assertequals(rawbytes, rawbytes2);
  }

  public void testparseextensions() throws exception {
    // testalltypes and testallextensions should have compatible wire formats,
    // so if we serialize a testalltypes then parse it as testallextensions
    // it should work.

    testalltypes message = testutil.getallset();
    bytestring rawbytes = message.tobytestring();

    extensionregistry registry = testutil.getextensionregistry();

    testallextensions message2 =
      testallextensions.parsefrom(rawbytes, registry);

    testutil.assertallextensionsset(message2);
  }

  public void testparsepackedextensions() throws exception {
    // ensure that packed extensions can be properly parsed.
    testpackedextensions message = testutil.getpackedextensionsset();
    bytestring rawbytes = message.tobytestring();

    extensionregistry registry = testutil.getextensionregistry();

    testpackedextensions message2 =
        testpackedextensions.parsefrom(rawbytes, registry);

    testutil.assertpackedextensionsset(message2);
  }

  public void testparseextensionslite() throws exception {
    // testalltypes and testallextensions should have compatible wire formats,
    // so if we serialize a testalltypes then parse it as testallextensions
    // it should work.

    testalltypes message = testutil.getallset();
    bytestring rawbytes = message.tobytestring();

    extensionregistrylite registry_lite = testutil.getextensionregistrylite();

    testallextensionslite message2 =
      testallextensionslite.parsefrom(rawbytes, registry_lite);

    testutil.assertallextensionsset(message2);

    // try again using a full extension registry.
    extensionregistry registry = testutil.getextensionregistry();

    testallextensionslite message3 =
      testallextensionslite.parsefrom(rawbytes, registry);

    testutil.assertallextensionsset(message3);
  }

  public void testparsepackedextensionslite() throws exception {
    // ensure that packed extensions can be properly parsed.
    testpackedextensionslite message = testutil.getlitepackedextensionsset();
    bytestring rawbytes = message.tobytestring();

    extensionregistrylite registry = testutil.getextensionregistrylite();

    testpackedextensionslite message2 =
        testpackedextensionslite.parsefrom(rawbytes, registry);

    testutil.assertpackedextensionsset(message2);
  }

  public void testextensionsserializedsize() throws exception {
    assertequals(testutil.getallset().getserializedsize(),
                 testutil.getallextensionsset().getserializedsize());
  }

  public void testserializedelimited() throws exception {
    bytearrayoutputstream output = new bytearrayoutputstream();
    testutil.getallset().writedelimitedto(output);
    output.write(12);
    testutil.getpackedset().writedelimitedto(output);
    output.write(34);

    bytearrayinputstream input = new bytearrayinputstream(output.tobytearray());

    testutil.assertallfieldsset(testalltypes.parsedelimitedfrom(input));
    assertequals(12, input.read());
    testutil.assertpackedfieldsset(testpackedtypes.parsedelimitedfrom(input));
    assertequals(34, input.read());
    assertequals(-1, input.read());

    // we're at eof, so parsing again should return null.
    asserttrue(testalltypes.parsedelimitedfrom(input) == null);
  }

  private void assertfieldsinorder(bytestring data) throws exception {
    codedinputstream input = data.newcodedinput();
    int previoustag = 0;

    while (true) {
      int tag = input.readtag();
      if (tag == 0) {
        break;
      }

      asserttrue(tag > previoustag);
      previoustag = tag;
      input.skipfield(tag);
    }
  }

  public void testinterleavedfieldsandextensions() throws exception {
    // tests that fields are written in order even when extension ranges
    // are interleaved with field numbers.
    bytestring data =
      testfieldorderings.newbuilder()
        .setmyint(1)
        .setmystring("foo")
        .setmyfloat(1.0f)
        .setextension(unittestproto.myextensionint, 23)
        .setextension(unittestproto.myextensionstring, "bar")
        .build().tobytestring();
    assertfieldsinorder(data);

    descriptors.descriptor descriptor = testfieldorderings.getdescriptor();
    bytestring dynamic_data =
      dynamicmessage.newbuilder(testfieldorderings.getdescriptor())
        .setfield(descriptor.findfieldbyname("my_int"), 1l)
        .setfield(descriptor.findfieldbyname("my_string"), "foo")
        .setfield(descriptor.findfieldbyname("my_float"), 1.0f)
        .setfield(unittestproto.myextensionint.getdescriptor(), 23)
        .setfield(unittestproto.myextensionstring.getdescriptor(), "bar")
        .build().tobytestring();
    assertfieldsinorder(dynamic_data);
  }

  private extensionregistry gettestfieldorderingsregistry() {
    extensionregistry result = extensionregistry.newinstance();
    result.add(unittestproto.myextensionint);
    result.add(unittestproto.myextensionstring);
    return result;
  }

  public void testparsemultipleextensionranges() throws exception {
    // make sure we can parse a message that contains multiple extensions
    // ranges.
    testfieldorderings source =
      testfieldorderings.newbuilder()
        .setmyint(1)
        .setmystring("foo")
        .setmyfloat(1.0f)
        .setextension(unittestproto.myextensionint, 23)
        .setextension(unittestproto.myextensionstring, "bar")
        .build();
    testfieldorderings dest =
      testfieldorderings.parsefrom(source.tobytestring(),
                                   gettestfieldorderingsregistry());
    assertequals(source, dest);
  }

  public void testparsemultipleextensionrangesdynamic() throws exception {
    // same as above except with dynamicmessage.
    descriptors.descriptor descriptor = testfieldorderings.getdescriptor();
    dynamicmessage source =
      dynamicmessage.newbuilder(testfieldorderings.getdescriptor())
        .setfield(descriptor.findfieldbyname("my_int"), 1l)
        .setfield(descriptor.findfieldbyname("my_string"), "foo")
        .setfield(descriptor.findfieldbyname("my_float"), 1.0f)
        .setfield(unittestproto.myextensionint.getdescriptor(), 23)
        .setfield(unittestproto.myextensionstring.getdescriptor(), "bar")
        .build();
    dynamicmessage dest =
      dynamicmessage.parsefrom(descriptor, source.tobytestring(),
                               gettestfieldorderingsregistry());
    assertequals(source, dest);
  }

  private static final int unknown_type_id = 1550055;
  private static final int type_id_1 =
    testmessagesetextension1.getdescriptor().getextensions().get(0).getnumber();
  private static final int type_id_2 =
    testmessagesetextension2.getdescriptor().getextensions().get(0).getnumber();

  public void testserializemessageseteagerly() throws exception {
    testserializemessagesetwithflag(true);
  }

  public void testserializemessagesetnoteagerly() throws exception {
    testserializemessagesetwithflag(false);
  }

  private void testserializemessagesetwithflag(boolean eagerparsing)
      throws exception {
    extensionregistrylite.seteagerlyparsemessagesets(eagerparsing);
    // set up a testmessageset with two known messages and an unknown one.
    testmessageset messageset =
      testmessageset.newbuilder()
        .setextension(
          testmessagesetextension1.messagesetextension,
          testmessagesetextension1.newbuilder().seti(123).build())
        .setextension(
          testmessagesetextension2.messagesetextension,
          testmessagesetextension2.newbuilder().setstr("foo").build())
        .setunknownfields(
          unknownfieldset.newbuilder()
            .addfield(unknown_type_id,
              unknownfieldset.field.newbuilder()
                .addlengthdelimited(bytestring.copyfromutf8("bar"))
                .build())
            .build())
        .build();

    bytestring data = messageset.tobytestring();

    // parse back using rawmessageset and check the contents.
    rawmessageset raw = rawmessageset.parsefrom(data);

    asserttrue(raw.getunknownfields().asmap().isempty());

    assertequals(3, raw.getitemcount());
    assertequals(type_id_1, raw.getitem(0).gettypeid());
    assertequals(type_id_2, raw.getitem(1).gettypeid());
    assertequals(unknown_type_id, raw.getitem(2).gettypeid());

    testmessagesetextension1 message1 =
      testmessagesetextension1.parsefrom(
        raw.getitem(0).getmessage().tobytearray());
    assertequals(123, message1.geti());

    testmessagesetextension2 message2 =
      testmessagesetextension2.parsefrom(
        raw.getitem(1).getmessage().tobytearray());
    assertequals("foo", message2.getstr());

    assertequals("bar", raw.getitem(2).getmessage().tostringutf8());
  }

  public void testparsemessageseteagerly() throws exception {
    testparsemessagesetwithflag(true);
  }

  public void testparsemessagesetnoteagerly()throws exception {
    testparsemessagesetwithflag(false);
  }

  private void testparsemessagesetwithflag(boolean eagerparsing)
      throws exception {
    extensionregistrylite.seteagerlyparsemessagesets(eagerparsing);
    extensionregistry extensionregistry = extensionregistry.newinstance();
    extensionregistry.add(testmessagesetextension1.messagesetextension);
    extensionregistry.add(testmessagesetextension2.messagesetextension);

    // set up a rawmessageset with two known messages and an unknown one.
    rawmessageset raw =
      rawmessageset.newbuilder()
        .additem(
          rawmessageset.item.newbuilder()
            .settypeid(type_id_1)
            .setmessage(
              testmessagesetextension1.newbuilder()
                .seti(123)
                .build().tobytestring())
            .build())
        .additem(
          rawmessageset.item.newbuilder()
            .settypeid(type_id_2)
            .setmessage(
              testmessagesetextension2.newbuilder()
                .setstr("foo")
                .build().tobytestring())
            .build())
        .additem(
          rawmessageset.item.newbuilder()
            .settypeid(unknown_type_id)
            .setmessage(bytestring.copyfromutf8("bar"))
            .build())
        .build();

    bytestring data = raw.tobytestring();

    // parse as a testmessageset and check the contents.
    testmessageset messageset =
      testmessageset.parsefrom(data, extensionregistry);

    assertequals(123, messageset.getextension(
      testmessagesetextension1.messagesetextension).geti());
    assertequals("foo", messageset.getextension(
      testmessagesetextension2.messagesetextension).getstr());

    // check for unknown field with type length_delimited,
    //   number unknown_type_id, and contents "bar".
    unknownfieldset unknownfields = messageset.getunknownfields();
    assertequals(1, unknownfields.asmap().size());
    asserttrue(unknownfields.hasfield(unknown_type_id));

    unknownfieldset.field field = unknownfields.getfield(unknown_type_id);
    assertequals(1, field.getlengthdelimitedlist().size());
    assertequals("bar", field.getlengthdelimitedlist().get(0).tostringutf8());
  }

  public void testparsemessagesetextensioneagerly() throws exception {
    testparsemessagesetextensionwithflag(true);
  }

  public void testparsemessagesetextensionnoteagerly() throws exception {
    testparsemessagesetextensionwithflag(false);
  }

  private void testparsemessagesetextensionwithflag(boolean eagerparsing)
      throws exception {
    extensionregistrylite.seteagerlyparsemessagesets(eagerparsing);
    extensionregistry extensionregistry = extensionregistry.newinstance();
    extensionregistry.add(testmessagesetextension1.messagesetextension);

    // set up a rawmessageset with a known messages.
    int type_id_1 =
        testmessagesetextension1
            .getdescriptor().getextensions().get(0).getnumber();
    rawmessageset raw =
      rawmessageset.newbuilder()
        .additem(
          rawmessageset.item.newbuilder()
            .settypeid(type_id_1)
            .setmessage(
              testmessagesetextension1.newbuilder()
                .seti(123)
                .build().tobytestring())
            .build())
        .build();

    bytestring data = raw.tobytestring();

    // parse as a testmessageset and check the contents.
    testmessageset messageset =
        testmessageset.parsefrom(data, extensionregistry);
    assertequals(123, messageset.getextension(
        testmessagesetextension1.messagesetextension).geti());
  }

  public void testmergelazymessagesetextensioneagerly() throws exception {
    testmergelazymessagesetextensionwithflag(true);
  }

  public void testmergelazymessagesetextensionnoteagerly() throws exception {
    testmergelazymessagesetextensionwithflag(false);
  }

  private void testmergelazymessagesetextensionwithflag(boolean eagerparsing)
      throws exception {
    extensionregistrylite.seteagerlyparsemessagesets(eagerparsing);
    extensionregistry extensionregistry = extensionregistry.newinstance();
    extensionregistry.add(testmessagesetextension1.messagesetextension);

    // set up a rawmessageset with a known messages.
    int type_id_1 =
        testmessagesetextension1
            .getdescriptor().getextensions().get(0).getnumber();
    rawmessageset raw =
      rawmessageset.newbuilder()
        .additem(
          rawmessageset.item.newbuilder()
            .settypeid(type_id_1)
            .setmessage(
              testmessagesetextension1.newbuilder()
                .seti(123)
                .build().tobytestring())
            .build())
        .build();

    bytestring data = raw.tobytestring();

    // parse as a testmessageset and store value into lazy field
    testmessageset messageset =
        testmessageset.parsefrom(data, extensionregistry);
    // merge lazy field check the contents.
    messageset =
        messageset.tobuilder().mergefrom(data, extensionregistry).build();
    assertequals(123, messageset.getextension(
        testmessagesetextension1.messagesetextension).geti());
  }

  public void testmergemessagesetextensioneagerly() throws exception {
    testmergemessagesetextensionwithflag(true);
  }

  public void testmergemessagesetextensionnoteagerly() throws exception {
    testmergemessagesetextensionwithflag(false);
  }

  private void testmergemessagesetextensionwithflag(boolean eagerparsing)
      throws exception {
    extensionregistrylite.seteagerlyparsemessagesets(eagerparsing);
    extensionregistry extensionregistry = extensionregistry.newinstance();
    extensionregistry.add(testmessagesetextension1.messagesetextension);

    // set up a rawmessageset with a known messages.
    int type_id_1 =
        testmessagesetextension1
            .getdescriptor().getextensions().get(0).getnumber();
    rawmessageset raw =
      rawmessageset.newbuilder()
        .additem(
          rawmessageset.item.newbuilder()
            .settypeid(type_id_1)
            .setmessage(
              testmessagesetextension1.newbuilder()
                .seti(123)
                .build().tobytestring())
            .build())
        .build();

    // serialize rawmessageset unnormally (message value before type id)
    bytestring.codedbuilder out = bytestring.newcodedbuilder(
        raw.getserializedsize());
    codedoutputstream output = out.getcodedoutput();
    list<rawmessageset.item> items = raw.getitemlist();
    for (int i = 0; i < items.size(); i++) {
      rawmessageset.item item = items.get(i);
      output.writetag(1, wireformat.wiretype_start_group);
      output.writebytes(3, item.getmessage());
      output.writeint32(2, item.gettypeid());
      output.writetag(1, wireformat.wiretype_end_group);
    }
    bytestring data = out.build();

    // merge bytes into testmessageset and check the contents.
    testmessageset messageset =
        testmessageset.newbuilder().mergefrom(data, extensionregistry).build();
    assertequals(123, messageset.getextension(
        testmessagesetextension1.messagesetextension).geti());
  }
}
